components/locid/examples/canonicalize_locales.rs

```diff
@@ -0,0 +1,53 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+// A sample application which takes a comma separated list of locales,
+// makes them canonical and serializes the list back into a comma separated list.
+
+use std::env;
+
+use icu_locid::Locale;
+
+const DEFAULT_INPUT: &str = "sr-cyrL-rS, es-mx, und-arab-u-ca-Buddhist";
+
+fn canonicalize_locales(input: &str) -> String {
+    // 1. Parse the input string into a list of locales.
+    let locales: Vec<Locale> = input
+        .split(',')
+        .filter_map(|s| s.trim().parse().ok())
+        .collect();
+
+    // 2. Canonicalize and serialize the output.
+    let canonical_locales: Vec<String> = locales
+        .into_iter()
+        .filter_map(|locale| icu_locid::Locale::canonicalize(locale.to_string()).ok())
```

Comment:
you already included `Locale`, so no need to prefix it.

Reply:
So, this is a bit weird sequence here. `canonicalize` does parse/stringify pair taking a &str and returning a String.

What you are doing is taking &str, parsing to Locale, then serializing then canonicalizing.

It feels like if you want to test `canonicalize` you could skip step 1 and just iterate over split string?

Or you could make it step 1 `parse`, step 2 `serialize`, without the need to `canonicalize`, because serialized Locale is canonicalized.

Reply:
Good input - it did feel to me as overkill.

I took you first option of just canonicalizing the string input.

Reply:
Btw. I think there is a bug in parsing/canonicalization code when you have more than 1 -u extension.
Look at the input/output:
`
Input: es-mx-u-ca-islamic-co-search-nu-roman
Output: es-MX-u-ca-islamicco-searchnu-roman
`
It seems like code is missing - in key/value join.

Reply:
thank you! Filed as https://github.com/unicode-org/icu4x/issues/368